xbmc/windowing/wayland/WinSystemWayland.cpp

```diff
@@ -594,74 +790,125 @@ bool CWinSystemWayland::ResetSurfaceSize(std::int32_t width, std::int32_t height
-    CDisplaySettings::GetInstance().AddResolutionInfo(newResInfo);
-    CDisplaySettings::GetInstance().ApplyCalibrations();
-    switchToRes = static_cast<RESOLUTION> (CDisplaySettings::GetInstance().ResolutionInfoSize() - 1);
-  }
-
-  // RES_DESKTOP does not change usually, it is still the current resolution
-  // of the selected output
-
-  assert(switchToRes != RES_INVALID);
+    // Find matching Kodi resolution member
+    if (!m_bFullScreen)
+    {
+      switchToRes = RES_WINDOW;
+      SetWindowResolution(m_bufferSize.Width(), m_bufferSize.Height());
+      // Save window size: Use unscaled size so we can start with same size
+      // no matter what monitor we are placed on at start
+      CServiceBroker::GetSettings().SetInt(CSettings::SETTING_WINDOW_WIDTH, m_surfaceSize.Width());
+      CServiceBroker::GetSettings().SetInt(CSettings::SETTING_WINDOW_HEIGHT, m_surfaceSize.Height());
+      CServiceBroker::GetSettings().Save();
+    }
+    else
+    {
+      switchToRes = FindMatchingCustomResolution(m_bufferSize, m_fRefreshRate);
+      if (switchToRes == RES_INVALID)
+      {
+        // Add new resolution if none found
+        RESOLUTION_INFO newResInfo;
+        UpdateDesktopResolution(newResInfo, 0, m_bufferSize.Width(), m_bufferSize.Height(), m_fRefreshRate);
+        newResInfo.strOutput = CDisplaySettings::GetInstance().GetCurrentResolutionInfo().strOutput; // we just assume the compositor put us on the right output
+        CDisplaySettings::GetInstance().AddResolutionInfo(newResInfo);
+        CDisplaySettings::GetInstance().ApplyCalibrations();
+        switchToRes = static_cast<RESOLUTION> (CDisplaySettings::GetInstance().ResolutionInfoSize() - 1);
+      }
+    }
 
-  // Mark resolution so that we know it came from configure
-  CDisplaySettings::GetInstance().GetResolutionInfo(switchToRes).strId = CONFIGURE_RES_ID;
+    // RES_DESKTOP does not change usually, it is still the current resolution
+    // of the selected output
 
-  CSingleExit exit(g_graphicsContext);
+    assert(switchToRes != RES_INVALID);
 
-  // Force resolution update
-  // SetVideoResolution() automatically delegates to main thread via internal
-  // message if called from other threads
-  // This will call SetFullScreen() with the new resolution, which also updates
-  // the size of the egl_window etc. from m_nWidth/m_nHeight.
-  // The call always blocks, so the configuration lock must be released beforehand.
-  // FIXME Ideally this class would be completely decoupled from g_graphicsContext,
-  // but this is not possible at the moment before the refactoring is done.
-  g_graphicsContext.SetVideoResolution(switchToRes, true);
+    // Force resolution update without going through SetFullScreen etc. since
+    // we take care of size update here
+    // FIXME Ideally this class would be completely decoupled from g_graphicsContext,
+    // but this is not possible at the moment before some refactoring is done.
```

Comment:
I'm all open for a better solution, but what is there to possibly postpone? This issue does not affect only windowed mode. If we don't do this, full screen mode also does not work, which means Wayland is not usable.